src/transfers.ts

```diff
@@ -1,36 +1,164 @@
+  await (await db)
+    .defaults({ errors: [] as BlockProcessingError[] })
+    .get('errors')
+    .push({ blockNumber: faultBlockNumber, reason })
+    .write();
+
+  await (await db)
+    .set('lastBlockProcessed', faultBlockNumber)
+    .write();
+
+  console.log('Critical error, extiting...');
+  console.log('Faulty block:', faultBlockNumber);
+  console.log('Reason:', reason);
+  process.exit();
+}
+
+async function processBlock(api: ApiPromise, head: Header) {
+  try {
+    await new Promise(async (resolve, reject) => {
+      // Set block processing timeout to avoid infinite lock
+      const processingTimeout = setTimeout(
+        () => reject('Block processing timeout'),
+        BLOCK_PROCESSING_TIMEOUT
+      );
+      // Set lock to avoid processing multiple blocks at the same time
+      processingLock.lock(async () => {
+        const blockNumber = head.number;
+        const { lastBlockProcessed = FIRST_BLOCK_TO_PROCESS - 1 } = (await db).valueOf();
+
+        // Ignore blocks that are (or should be) already processed
+        if (blockNumber.toNumber() <= lastBlockProcessed) {
+          processingLock.unlock();
+          clearTimeout(processingTimeout);
+          return;
+        }
+
+        console.log(`\nProcessing block #${ blockNumber }...`);
+
+        const blockHash = head.hash;
+        const events = await api.query.system.events.at(blockHash) as Vec<EventRecord>;
+        const { sizeDollarPool: currentDollarPool = 0, tokensBurned: currentTokensBurned = 0 } = (await db).valueOf();
+        let sumDollarsInBlock = 0, sumTokensInBlock = 0;
+
+        // To calcultate the price we use parent hash (so all the transactions that happend in this block have no effect on it)
+        const price = await joy.price(head.parentHash, currentDollarPool);
+
+        // Processing all events in the finalized block
+        for (let { event } of events) {
+          if (event.section === 'balances' && event.method === 'Transfer') {
+            const recipient = event.data[1] as AccountId;
+            if (recipient.toString() === process.env.JSGENESIS_ADDRESS) {
+              // For all events of "Transfer" type with matching recipient...
+              const sender = event.data[0] as AccountId;
+              const amountJOY = event.data[2] as Balance;
+              const feesJOY = event.data[3] as Balance;
+              const timestamp = await api.query.timestamp.now.at(blockHash) as Moment;
+              const memo = await api.query.memo.memo.at(blockHash, sender) as Text;
+
+              const exchange: Exchange = {
+                sender: sender.toString(),
+                recipient: recipient.toString(),
+                xmrAddress: memo.toString(),
+                amount: amountJOY.toNumber(),
+                fees: feesJOY.toNumber(),
+                date: new Date(timestamp.toNumber()),
+                blockHeight: blockNumber.toNumber(),
+                price: parseFloat(price),
```

Comment:
It would make sense to add the `totalUSD` here too so we don't have to manually calculate it on the website or somewhere else

Reply:
Added this as `amountUSD` (since we already had `amount` for tokens)